nixos/modules/config/ipa.nix

```
@@ -0,0 +1,241 @@
+        description = "Whether to whitelist the FreeIPA domain in Chromium.";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = sssd.enable;
+        message = "sssd must be enabled through `services.sssd.enable` for FreeIPA integration to work.";
+      }
+      {
+        assertion = ntp.enable;
+        message = "ntpd must be enabled through `services.ntp.enable` for FreeIPA integration to work.";
+      }
+      {
+        assertion = !krb5.enable;
+        message = "krb5 must be disabled through `krb5.enable` for FreeIPA integration to work.";
+      }
+      {
+        assertion = !ldap.enable;
+        message = "ldap must be disabled through `users.ldap.enable` for FreeIPA integration to work.";
+      }
+    ];
+
+    environment.systemPackages = with pkgs; [ krb5Full freeipa ];
+
+    environment.etc = mkMerge [{
+      "ipa/default.conf".text = ''
+        [global]
+        basedn = ${cfg.basedn}
+        realm = ${cfg.realm}
+        domain = ${cfg.domain}
+        server = ${cfg.server}
+        host = ${config.networking.hostName}
+        xmlrpc_uri = https://${cfg.server}/ipa/xml
+        enable_ra = True
+      '';
+
+      "ipa/nssdb".source = nssDb;
+
+      "krb5.conf".text = ''
+        [libdefaults]
+         default_realm = ${cfg.realm}
+         dns_lookup_realm = false
+         dns_lookup_kdc = true
+         rdns = false
+         ticket_lifetime = 24h
+         forwardable = true
+         udp_preference_limit = 0
+
+        [realms]
+         ${cfg.realm} = {
+          kdc = ${cfg.server}:88
+          master_kdc = ${cfg.server}:88
+          admin_server = ${cfg.server}:749
+          default_domain = ${cfg.domain}
+          pkinit_anchors = FILE:/etc/ipa/ca.crt
+        }
+
+        [domain_realm]
+         .${cfg.domain} = ${cfg.realm}
+         ${cfg.domain} = ${cfg.realm}
+         ${cfg.server} = ${cfg.realm}
+
+        [dbmodules]
+          ${cfg.realm} = {
+            db_library = ${pkgs.freeipa}/lib/krb5/plugins/kdb/ipadb.so
+          }
+      '';
+
+      "opanldap/ldap.conf".source = ldapConf;
+    }
+    (mkIf cfg.chromiumSupport {
+      "chromium/policies/managed/freeipa.json".text = ''
+        { "AuthServerWhitelist": "*.${cfg.domain}" }
+      '';
+    })];
+
+    system.activationScripts.ipa = stringAfter [ "etc" ] ''
+      # libcurl requires a hard copy of the certificate
+      if ! ${pkgs.diffutils}/bin/diff ${cfg.certificate} /etc/ipa/ca.crt > /dev/null 2>&1; then
+        rm -f /etc/ipa/ca.crt
+        cp ${cfg.certificate} /etc/ipa/ca.crt
+      fi
+
+      if [ ! -f /etc/krb5.keytab ]; then
+        cat <<EOF
+
+In order to complete FreeIPA integration, please join the domain by completing the following steps:
+1. Authenticate as an IPA user authorized to join new hosts, e.g. kinit admin@${cfg.realm}
+2. Join the domain and obtain the keytab file: ipa-join
+3. Install the keytab file: sudo install -m 600 krb5.keytab /etc/
+
+EOF
```

If you use nixpkgs-fmt then this will be intended

```suggestion
      In order to complete FreeIPA integration, please join the domain by completing the following steps:
      1. Authenticate as an IPA user authorized to join new hosts, e.g. kinit admin@${cfg.realm}
      2. Join the domain and obtain the keytab file: ipa-join
      3. Install the keytab file: sudo install -m 600 krb5.keytab /etc/
EOF
```